custom_components/tahoma/__init__.py

```diff
@@ -36,52 +35,59 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
-    except RequestException:
-        _LOGGER.exception("Error when getting devices from the TaHoma API")
+        client = TahomaClient(username, password)
+        await client.login()
+    except TooManyRequestsException as exception:
+        _LOGGER.exception(exception)
         return False
+    except BadCredentialsException as exception:
+        _LOGGER.exception(exception)
+        return False
+    except Exception as exception:  # pylint: disable=broad-except
+        _LOGGER.exception(exception)
+        return False
+
+    devices = await client.get_devices()
+    scenes = await client.get_scenarios()
+
+    hass.data[DOMAIN][entry.entry_id] = {"controller": client, "entities": {}}
 
-    hass.data[DOMAIN][entry.entry_id] = {
-        "controller": controller,
-        "devices": [],
-        "scenes": [],
-    }
+    for platform in PLATFORMS:
+        hass.data[DOMAIN][entry.entry_id]["entities"][platform] = []
+
+    for scene in scenes:
+        hass.data[DOMAIN][entry.entry_id]["entities"]["scene"].append(scene)
 
     for device in devices:
-        _device = controller.get_device(device)
+        if device.widget in TAHOMA_TYPES or device.ui_class in TAHOMA_TYPES:
+            platform = TAHOMA_TYPES.get(device.widget) or TAHOMA_TYPES.get(
+                device.ui_class
+            )
 
-        if _device.uiclass in TAHOMA_TYPES:
-            if TAHOMA_TYPES[_device.uiclass] in PLATFORMS:
-                component = TAHOMA_TYPES[_device.uiclass]
-                hass.data[DOMAIN][entry.entry_id]["devices"].append(_device)
+            if platform in PLATFORMS:
+                hass.data[DOMAIN][entry.entry_id]["entities"][platform].append(device)
 
-        elif _device.type not in [
-            "ogp:Bridge"
+        elif device.controllable_name not in [
+            "ogp:Bridge",
+            "internal:PodV2Component",
+            "internal:TSKAlarmComponent",
         ]:  # Add here devices to hide from the debug log.
             _LOGGER.debug(
-                "Unsupported Tahoma device (%s). Create an issue on Github with the following information. \n\n %s \n %s \n %s",
-                _device.type,
-                _device.type + " - " + _device.uiclass + " - " + _device.widget,
-                json.dumps(_device.command_def) + ",",
-                json.dumps(_device.states_def),
+                "Unsupported Tahoma device (%s - %s - %s). Create an issue on Github to discuss the possibilities.",
+                device.controllable_name,
+                device.ui_class,
+                device.widget,
             )
 
-    for component in PLATFORMS:
-        hass.async_create_task(
-            hass.config_entries.async_forward_entry_setup(entry, component)
-        )
-
-    for scene in scenes:
-        hass.data[DOMAIN][entry.entry_id]["scenes"].append(scene)
+    for platform in hass.data[DOMAIN][entry.entry_id]["entities"]:
+        if len(hass.data[DOMAIN][entry.entry_id]["entities"][platform]) > 0:
```

Comment:
Now only platforms with existing entities will be set up. This makes our integration more efficient. I am not a fan of the if statement / loop structure, looking for suggestions.

Reply:
Look at default dict:

```python
>>> from collections import defaultdict
>>> d = defaultdict(list)
>>>d["foo"].append("faa")
``

Reply:
@tetienne defaultdict is implemented! I still don't really like this big loop, any idea how to make this nicer?

Reply:
I think you have no choice here. You have to trigger the entry setup for all platform.